### PR TITLE
Performance improvement: cache app classloaders

### PR DESCRIPTION
**Pull request by tudor-malene**
First Corda PR.

Reason:
- when running unit tests I noticed tests scanning the classpath repeatedly, which is usually very slow

Fix:
- I've added a cache, and it reduced the "node:test" time by about 15-20s (locally)  
- as a cache implementation I used a LRUMap with a max size to avoid memory issues

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/internal/cordapp/CordappLoader.kt [hunk 1]
```diff
@@ -67,6 +68,9 @@ class CordappLoader private constructor(private val cordappJarPaths: List<Restri
          */
         fun createDefault(baseDir: Path) = CordappLoader(getCordappsInDirectory(getCordappsPath(baseDir)))
 
+        // Cache for CordappLoaders to avoid costly classpath scanning
+        private val cordappLoadersCache = LRUMap<Any, CordappLoader>(1000)
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/cordapp/CordappLoader.kt]: Any? Can't we use a more specific type here?

Comment (tudor-malene) [reply node/src/main/kotlin/net/corda/node/internal/cordapp/CordappLoader.kt]: It's used by 2 functions, one uses List<String>, and the other List<RestrictedURL> to find the packages to scan.
It didn't feel like the type of the key was really relevant for this.

Another option, if we want to type the key, is to have have one of these Caches per key type.
Or maybe List<*> ?
Or add some complexity to those methods to convert to List<String>, but that wouldn't have a lot of benefit

Comment (mikehearn) [reply node/src/main/kotlin/net/corda/node/internal/cordapp/CordappLoader.kt]: Yeah `List<*>` sounds like the right type then.

Comment (tudor-malene) [reply node/src/main/kotlin/net/corda/node/internal/cordapp/CordappLoader.kt]: Cool.

Regarding test mode detection: Should I investigate how to do that?

Comment (mikehearn) [reply node/src/main/kotlin/net/corda/node/internal/cordapp/CordappLoader.kt]: It's not that important so don't spend too much time on it - maybe Clinton has an idea. I'm sure there must be a way to find out if JUnit is starting things up for instance. The node has a notion of integration test mode already, I think.

Comment (tudor-malene) [reply node/src/main/kotlin/net/corda/node/internal/cordapp/CordappLoader.kt]: I can't see an easy way. 
Maybe passing some parameter.. 

These methods have "test" in their name anyway, which should be discouraging to use from production code.

**Pull request merged**